File: riistaandroid/src/main/java/fi/riista/mobile/activity/LoginActivity.java
package fi.riista.mobile.activity;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.Menu;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.widget.Button;
import android.widget.EditText;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Locale;

import fi.riista.mobile.R;
import fi.riista.mobile.database.GameDatabase;
import fi.riista.mobile.models.announcement.Announcement;
import fi.riista.mobile.network.LoginTask;
import fi.riista.mobile.storage.StorageDatabase;
import fi.riista.mobile.utils.AppPreferences;
import fi.riista.mobile.utils.JsonUtils;
import fi.riista.mobile.utils.Utils;
import fi.vincit.androidutilslib.activity.WorkActivity;

/**
 * Login screen is skipped if user has stored credentials
 */
public class LoginActivity extends WorkActivity {

    public static final String EVENT_YEARS_EXTRA = "gameDiaryYears";
    public static final String SHOW_ANNOUNCEMENTS_EXTRA = "showAnnouncements";

    private Button mLoginButton = null;
    private boolean showAnnouncements = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        checkNotificationIntentData();

        setupLocaleFromSetting();
        setContentView(R.layout.activity_loginscreen);
        setupButtons(false);
    }

    private void checkNotificationIntentData() {
        String announcement = getIntent().getStringExtra("announcement");
        if (announcement != null) {
            //We got announcement data from a notification, so save it if we can.
            showAnnouncements = true;
            Announcement ann = JsonUtils.jsonToObject(announcement, Announcement.class);
            if (ann != null) {
                StorageDatabase.getInstance().updateAnnouncement(ann, null);
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        boolean googleServicesEnabled = Utils.checkPlayServices(this);
        mLoginButton.setEnabled(googleServicesEnabled);
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        setupButtons(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.loginscreen, menu);
        return true;
    }

    public void attemptLogin(final String username, final String password) {
        LoginTask log = new LoginTask(getWorkContext(), username, password, false) {
            @Override
            public void loginFailed() {
                mLoginButton.setEnabled(true);
            }

            @Override
            public void onFinishText(String text) {
                GameDatabase.getInstance().storeCredentials(LoginActivity.this, username, password);
                Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                try {
                    JSONObject jObject = new JSONObject(text);
                    JSONArray yearArray = jObject.getJSONArray("gameDiaryYears");
                    ArrayList<Integer> eventYears = new ArrayList<>();
                    for (int i = 0; i < yearArray.length(); i++) {
                        eventYears.add(yearArray.getInt(i));
                    }
                    intent.putIntegerArrayListExtra(EVENT_YEARS_EXTRA, eventYears);

                    AppPreferences.setUserInfo(getWorkContext().getContext(), text);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                replaceWithMainActivity(intent);
            }
        };
        log.start();
        mLoginButton.setEnabled(false);
    }

    public void setupLocaleFromSetting() {
        String languageSelected = AppPreferences.getLanguageCodeSetting(this);
        Locale locale = new Locale(languageSelected);
        Locale.setDefault(locale);
        Configuration config = new Configuration();
        config.locale = locale;
        getBaseContext().getResources().updateConfiguration(config, getBaseContext().getResources().getDisplayMetrics());
    }

    void setupButtons(boolean restarted) {
        mLoginButton = (Button) findViewById(R.id.loginbutton);
        if (Utils.checkPlayServices(this)) {

            // Login automatically if credentials are stored
            if (GameDatabase.getInstance().credentialsStored(this)) {
                if (restarted) {
                    finish();
                    return;
                }
                GameDatabase.getInstance().loginWithStoredCredentials(getWorkContext(), false);
                Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                replaceWithMainActivity(intent);
                return;
            }

            mLoginButton.setEnabled(true);
            mLoginButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    EditText usernameField = (EditText) findViewById(R.id.username);
                    EditText passwordField = (EditText) findViewById(R.id.password);
                    attemptLogin(usernameField.getText().toString(), passwordField.getText().toString());
                }
            });
        }
        findViewById(R.id.layout).setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent ev) {
                Utils.hideKeyboard(LoginActivity.this, view);
                return false;
            }
        });
    }

    void replaceWithMainActivity(Intent intent) {
        //Check if the main activity should show the announcement page
        boolean show = getIntent().getBooleanExtra(SHOW_ANNOUNCEMENTS_EXTRA, false) || showAnnouncements;
        intent.putExtra(SHOW_ANNOUNCEMENTS_EXTRA, show);
        startActivity(intent);

        finish();
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/activity/LoginActivity.kt
package fi.riista.mobile.activity

import android.content.DialogInterface
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.EditText
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import dagger.android.AndroidInjection
import fi.riista.mobile.R
import fi.riista.mobile.RiistaApplication
import fi.riista.mobile.models.announcement.Announcement
import fi.riista.mobile.network.CheckVersionTask
import fi.riista.mobile.storage.StorageDatabase
import fi.riista.mobile.ui.UpdateAvailableDialog
import fi.riista.mobile.utils.*
import fi.riista.mobile.utils.Authenticator.AuthCallback
import fi.riista.mobile.utils.Authenticator.AuthSuccessCallback
import fi.riista.mobile.utils.LocaleUtil.localeFromLanguageSetting
import fi.vincit.androidutilslib.activity.WorkActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference
import java.util.*
import javax.inject.Inject

/**
 * Login screen is skipped if user has stored credentials
 */
class LoginActivity : WorkActivity() {
    @Inject
    lateinit var authenticator: Authenticator

    @Inject
    lateinit var credentialsStore: CredentialsStore

    private lateinit var mUsernameInput: TextInputEditText
    private lateinit var mPasswordInput: TextInputEditText
    private lateinit var mLoginButton: MaterialButton

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setupLocaleFromSetting()
        setContentView(R.layout.activity_loginscreen)
        mUsernameInput = findViewById(R.id.username)
        mPasswordInput = findViewById(R.id.password)
        mLoginButton = findViewById(R.id.loginbutton)

        val activityLaunchedByAnnouncementNotification = extractAnnouncementFromNotificationIntentData()
        setupButtons(activityLaunchedByAnnouncementNotification)
    }

    override fun onResume() {
        super.onResume()
        mLoginButton.isEnabled = Utils.isPlayServicesAvailable(this, true)
        if (!RiistaApplication.sDidCheckVersion) {
            RiistaApplication.sDidCheckVersion = true
            checkVersionUpdate()
        }
    }

    // This method is relevant in case a message with both notification and data payload is
    // received while application is not in the foreground (either not active or not running at
    // all). The notification is then delivered to the device’s system tray, and the data payload
    // is delivered in the extras of the intent of your launcher Activity.
    //
    // Returns true if announcement was present in notification intent.
    private fun extractAnnouncementFromNotificationIntentData(): Boolean {
        val announcement = intent.getStringExtra("announcement") ?: return false
        JsonUtils.jsonToObject(announcement, Announcement::class.java, true)
            ?.let {
                // We got announcement data from a notification, so save it if we can.
                StorageDatabase.getInstance().updateAnnouncement(it, null)
            }

        return true
    }

    private fun attemptLogin(username: String, password: String) {
        CoroutineScope(Main).launch {
            authenticator.authenticate(username, password, LoginCallback(this@LoginActivity))
        }
        mLoginButton.isEnabled = false
    }

    private fun setupLocaleFromSetting() {
        val locale = localeFromLanguageSetting(this)
        Locale.setDefault(locale)
        val config = Configuration()
        // TODO: 2.2.2021 fix deprecated locale usage
        config.locale = locale
        val res = baseContext.resources
        res.updateConfiguration(config, res.displayMetrics)
    }

    private fun setupButtons(isActivityStartedByAnnouncementNotification: Boolean) {
        if (Utils.isPlayServicesAvailable(this, true)) {

            // Login automatically if credentials are stored.
            if (credentialsStore.isCredentialsSaved()) {
                CoroutineScope(Main).launch {
                    authenticator.reauthenticate(
                            ReloginCallback(
                                    this@LoginActivity,
                                    isActivityStartedByAnnouncementNotification
                            )
                    )
                }
                val intent = Intent(this, MainActivity::class.java)

                // Sync is not enabled now since MainActivity will be triggered a second time
                // via onNewIntent method after login task has completed.
                intent.putExtra(
                        MainActivity.DO_NOT_INITIATE_SYNC,
                        isActivityStartedByAnnouncementNotification
                )
                startActivity(intent)
                if (!isActivityStartedByAnnouncementNotification) {
                    // In case of announcement notification finish() will be called when login
                    // task has completed.
                    finish()
                }
                return
            }
            mLoginButton.isEnabled = true
            mLoginButton.setOnClickListener {
                val username = findViewById<EditText>(R.id.username).text.toString()
                val password = findViewById<EditText>(R.id.password).text.toString()
                attemptLogin(username, password)
            }
        }
        findViewById<View>(R.id.layout).setOnTouchListener { view: View, ev: MotionEvent ->
            if (ev.action == MotionEvent.ACTION_UP) {
                view.performClick()
            }
            if (mUsernameInput.isFocused) {
                mUsernameInput.clearFocus()
            }
            if (mPasswordInput.isFocused) {
                mPasswordInput.clearFocus()
            }
            KeyboardUtils.hideKeyboard(this@LoginActivity, view)
            false
        }
    }

    private fun replaceActivity(intent: Intent) {
        startActivity(intent)
        finish()
    }

    private fun checkVersionUpdate() {
        Log.d(TAG, "Check version")
        val task: CheckVersionTask = object : CheckVersionTask(workContext) {
            override fun onFinishText(text: String) {
                Log.d(TAG, "Version check finished $text")
                RiistaApplication.sDidCheckVersion = true
                if (Utils.shouldDisplayVersionUpdateDialog(text)) {
                    UpdateAvailableDialog.show(this@LoginActivity, "", 0)
                }
            }

            override fun onError() {
                Log.d(TAG, "Version check error")
            }
        }
        task.start()
    }

    private class LoginCallback(activity: LoginActivity) : AuthCallback {
        private val mActivityReference: WeakReference<LoginActivity> = WeakReference(activity)

        override fun onLoginSuccessful(userInfo: String?) {
            mActivityReference.get()?.onLoginSuccessful()
        }

        override fun onLoginFailed(httpStatusCode: Int) {
            mActivityReference.get()?.takeIf { !it.isFinishing }?.onLoginFailed(httpStatusCode)
        }

    }

    private fun onLoginSuccessful() {
        val intent = Intent(this, MainActivity::class.java)
        replaceActivity(intent)
    }

    private fun onLoginFailed(httpStatusCode: Int) {
        @StringRes val msgResId: Int = when (httpStatusCode) {
            418 -> {
                R.string.version_outdated
            }
            403 -> {
                R.string.login_failed
            }
            else -> {
                R.string.connecting_failed
            }
        }
        val message = getString(msgResId)
        AlertDialog.Builder(this)
            .setMessage(message)
            .setPositiveButton(R.string.ok) { _: DialogInterface?, _: Int -> }
            .show()

        mLoginButton.isEnabled = true
    }

    private fun onReloginSuccessful() {
        val intent = Intent(this, MainActivity::class.java)
        intent.putExtra(SHOW_ANNOUNCEMENTS_EXTRA, true)

        // Instead of creating a new instance of MainActivity on top of current one,
        // onNewIntent method of the current instance will be called.
        intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
        replaceActivity(intent)
    }

    private class ReloginCallback(
        activity: LoginActivity,
        isActivityStartedByAnnouncementNotification: Boolean
    ) : AuthSuccessCallback() {
        private val mActivityReference: WeakReference<LoginActivity> = WeakReference(activity)
        private val mIsActivityStartedByAnnouncementNotification: Boolean =
            isActivityStartedByAnnouncementNotification

        override fun onLoginSuccessful(userInfo: String?) {
            mActivityReference.get()
                ?.takeIf { mIsActivityStartedByAnnouncementNotification }
                ?.onReloginSuccessful()
        }

    }

    companion object {
        const val SHOW_ANNOUNCEMENTS_EXTRA = "showAnnouncements"
        private const val TAG = "LoginActivity"
    }
}
